### ISSUE TITLE
Overriding a method with explicit void as return type with a method that has a different return type doesn't give an error

### ISSUE THREAD

**Issue by drake7707** [Microsoft/TypeScript#4544]
In the following code I expected the compiler to complain because I'm changing the return type of doSomething in the subclass from no return value to a method that returns a value. 

```
class Hello {
    doSomething():void {
        
    }
}
class SubHello extends Hello {   
    doSomething():boolean {
        return true;
    }
}
```
I understand that not explicitly specifying void in the super class would leave the return type open to interpretation in the derived classes (not sure if that's the intent), but explicitly saying it shouldn't return anything at all should enforce the no returning of anything in the derived classes.

I looked it up in the language spec to see if I could dig up anything of relevance to this issue:

The language spec defines the void type as follows:
>The Void type, referenced by the  void keyword, represents the absence of a value and is used as the return type of functions with no return value.
>The only possible values for the Void type are  null and  undefined . The Void type is a subtype of the Any type and a supertype of the Null and Undefined types, but otherwise Void is unrelated to all other types.

Furthermore:

>A property member in a derived class is said to override a property member in a base class when the derived class property member has the same name and kind (instance or static) as the base class property member. The type of an overriding property member must be assignable (section 3.10.4) to the type of the overridden property member, or otherwise a compile-time error occurs.

**Comment by jesseschalken** [comment]
This looks wrong to me, too.

The compiler properly enforces that `boolean` isn't compatible with `void` here:

```typescript
var foo:void = true;
// Type 'boolean' is not assignable to type 'void'.
```

But not when it's the return type in a function signature. For example, this code has the exact same error, but passes:

```typescript
var func:() => void = () => true;
var foo:void = func();
// OK
console.log(foo);
// true
```

I'm not sure why `void` return types get special treatment.

*(tsc 1.5.3)*

**Comment by vladima** [comment]
Current behavior is by design:

[The Void Type](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#325-the-void-type)
>The only possible values for the Void type are null and undefined. The Void type is a subtype of the Any type and a supertype of the Null and Undefined types, but otherwise Void is unrelated to all other types.

 [Simple Variable Declarations] (https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#5.2.1) 
>When a variable declaration specifies both a type annotation and an initializer expression, the type of the initializer expression is required to be assignable to (section 3.11.4) the type given in the type annotation.
```ts
let x: void = true; // error: boolean is not assignable to void
```

[Return Statements] (https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#510-return-statements)
>When a 'return' statement includes an expression, if the containing function includes a return type annotation, the return expression is contextually typed (section 4.23) by that return type and must be of a type that is assignable to the return type.
```ts
function y(): void {
    return true; // error: boolean is not assignable to void
}
```

The same rule also applied to arrow functions given that per [ArrowFunctions](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#411-arrow-functions):
>An arrow function expression of the form
```ts
( ... ) => expr
```
is exactly equivalent to
```ts
( ... ) => { return expr ; }
```

```ts
let z = (): void => true; // error: boolean is not assignable to void
```

This sample however is slightly different:
```ts
var func:() => void = () => true;
```

This is a variable declarations so we need to verify that initializer is assignable to type annotation.
 Type of initializer is `() => boolean` which is effectively `{ (): boolean }` and annotated type is `() => void` (`{(): void}`). Per [Assignment Compatibility](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3114-assignment-compatibility) they are indeed assignment compatible since call signature M in T (type annotation) and call signature N in S (initializer) both has no parameters and M has `void` as return type. 
AFAIR rationale for this is:
- it is safe to ignore a return value
- it simplifies quite common scenario when function value with non-void return type is passed as an argument to something that expects void-returning function otherwise it will require using boilerplate code like  `foo(() => void bar)` instead of just `foo(bar)` )
